rfc/datawherehouse.md

```diff
@@ -0,0 +1,236 @@
+
+## High level flow of proposed solution
+
+* Client stores some blob with web3.storage
+* Client requests this blob to be published with `store/publish`
+* Service handler verifies requested blob was stored on the client space, issues a location claim with encoded information for location and returns it to the client
+  * Can be batched with other invocations like `filecoin/offer`
+
+The following diagram presents the described flow, and is illustrated with the following steps:
+
+1. Client requests to store some bytes with the storefront service
+2. Service issues a receipt stating the URI of a write target for the client to write those bytes
+3. Client writes the bytes into the write target
+4. Client requests service to serve written bytes under given stored CID
+5. Service verifies that the bytes are stored in the provided write target/space and writes a claim about it.
+6. Service issues a receipt stating bytes are being stored by the service on a given location claim.
+
+![datawherehouse1](./datawherehouse/datawherehouse-1.svg)
+
+On the other side of things, the next diagram presents the flow when a client wants to read some blob, and is illustrated with the following steps:
+
+1. Client request to read some bytes by the CID of the blob
+2. Service discovers where the requested bytes are stored relying on content claims service to find location claim from the service.
+3. Service serves blob stored on the discovered location target.
+
+![datawherehouse2](./datawherehouse/datawherehouse-2.svg)
+
+## `store/publish` capability
+
+After some blob being stored with web3.storage, a client MAY request this blob to be available under given CID. Service MUST verify that it has bytes corresponding to that CID and it was stored on the provided client space. If so, service MUST issue a location claim to the content claims service. Moreover, the service SHOULD respond with location claim from which read can be performed.
+
+This method enables the service to handle private data in the future, and likely should allow client to specify TTL for the produced read URL, as well as even consider future where this read URI might require permission (e.g. ucan authorization).
+
+This capability can be specified as follows:
+
+```json
+{
+  "op": "store/publish",
+  "rsc": "did:key:abc...space",
+  "input": {
+    "link": { "/": "bafy...BLOBCID" }, // RAW CID - go from CAR <-> RAW CID it's just switching a codec byte
+    "url": "https://..." // write target presignedurl previously provided
+  }
+}
+```
+
+Return on success the following receipt:
+
+```json
+{
+  "ran": "bafy...storePublish",
+  "out": {
+    "ok": {
+      "link" : { "/": "bafy...BLOBCID" }, // RAW CID - go from CAR <-> RAW CID it's just switching a codec byte
+      "location": "`https://w3s.link/ipfs/bafy...BLOBCID?origin=r2://region/bucketName/key"
+    }
+  },
+  "fx": {},
+}
+```
+
+TODO: Should return the actual location claim? How would we do that?
```

Comment:
TODO: would love your help here @Gozala ?

Reply:
or maybe a CID for it? but client can't really read one today? Maybe receipt CID that claim was made?

Reply:
embed would mean we need to implement in ucanto what we talked about in Istanbul to attach things via receipt builder, but that seems good